src/audio/mixin_mixout.c

```diff
@@ -0,0 +1,999 @@
+	src = source->r_ptr;
+
+	assert(mixed_bytes >= start);
+	bytes_to_mix = MIN(mixed_bytes - start, size);
+
+	bytes_processed = 0;
+
+	while (bytes_processed < bytes_to_mix) {
+		n = MIN(audio_stream_bytes_without_wrap(sink, dest),
+			bytes_to_mix - bytes_processed);
+		n = MIN(audio_stream_bytes_without_wrap(source, src), n);
+		n >>= 2;	/* divide by 4 */
+		for (i = 0; i < n; i++) {
+			*dest = sat_int32((int64_t)*dest + (int64_t)*src);
+			src++;
+			dest++;
+		}
+
+		bytes_processed += n << 2;
+		dest = audio_stream_wrap(sink, dest);
+		src = audio_stream_wrap(source, src);
+	}
+
+	audio_stream_bytes_copy(sink, dest, source, src, size - bytes_to_mix);
+}
+#endif	// CONFIG_FORMAT_S32LE
+
+// mix silence into stream, i.e. set not yet mixed data in stream to zero
+static void silence(struct audio_stream __sparse_cache *stream, uint32_t start,
+		    uint32_t mixed_bytes, uint32_t size)
+{
+	uint32_t skip_mixed;
+	uint8_t *ptr;
+	int n;
+
+	assert(mixed_bytes >= start);
+	skip_mixed = mixed_bytes - start;
+
+	if (size <= skip_mixed)
+		return;
+
+	size -= skip_mixed;
+	ptr = audio_stream_wrap(stream, (uint8_t *)stream->w_ptr + mixed_bytes);
+
+	while (size) {
+		n = MIN(audio_stream_bytes_without_wrap(stream, ptr), size);
+		memset(ptr, 0, n);
+		size -= n;
+		ptr = audio_stream_wrap(stream, ptr + n);
+	}
+}
+
+static int mixin_copy(struct comp_dev *dev)
+{
+	struct mixin_data *mixin_data;
+	struct comp_buffer *source;
+	struct comp_buffer __sparse_cache *source_c;
+	uint32_t source_avail_bytes, sinks_free_bytes;
+	struct list_item *blist;
+	uint32_t bytes_to_consume_from_source_buf;
+	uint32_t bytes_to_copy;
+
+	comp_dbg(dev, "mixin_copy()");
+
+	mixin_data = comp_get_drvdata(dev);
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	source_c = buffer_acquire(source);
+
+	source_avail_bytes = audio_stream_get_avail_bytes(&source_c->stream);
+	sinks_free_bytes = INT32_MAX;
+
+	list_for_item(blist, &dev->bsink_list) {
+		struct comp_buffer *unused_in_between_buf;
+		struct comp_dev *mixout;
+		struct comp_buffer *sink;
+		struct mixout_data *mixout_data;
+		struct mixed_data_info __sparse_cache *mixed_data_info;
+		struct mixout_source_info *src_info;
+		struct comp_buffer __sparse_cache *sink_c;
+		uint32_t stream_free_bytes;
+
+		// unused buffer between mixin and mixout
+		unused_in_between_buf = buffer_from_list(blist, struct comp_buffer,
+							 PPL_DIR_DOWNSTREAM);
+		mixout = buffer_get_comp(unused_in_between_buf, PPL_DIR_DOWNSTREAM);
+		sink = list_first_item(&mixout->bsink_list, struct comp_buffer, source_list);
+
+		mixout_data = comp_get_drvdata(mixout);
+		mixed_data_info = mixed_data_info_acquire(mixout_data->mixed_data_info);
+
+		src_info = find_mixout_source_info(mixed_data_info, dev);
+		assert(src_info);
+
+		sink_c = buffer_acquire(sink);
+
+		stream_free_bytes = audio_stream_get_free_bytes(&sink_c->stream);
+		assert(stream_free_bytes >= src_info->consumed_yet_not_produced_bytes);
```

Comment:
Can this check be non fatal i.e. we should log() the error and return the error up the stack and allow other unrelated pipelines to continue.

Reply:
Actually that condition (stream_free_bytes >= src_info->consumed_yet_not_produced_bytes) is always true. The only possibility it will evaluate to false if someone modified the code and introduce a bug.

So I think it is better to either remove that assert() altogether or leave it as is. But replacing it with if (...) { log_err(...) } seems not a good idea -- the error would just indicate that someone introduced a bug in code and so it is better to signal it in a clear way instead of trying to run other pipelines possibly hiding the issue.

I chose assert() as in case of a bug the math will broke and uint32_t overflow will occur possibly resulting in writing beyond buffer allocated memory which might result in magic issues hard to find.

This assert() is inside the loop, however, this loop iterates by mixouts connected to single mixing. Most often it is just 1 with max 3 (for Meteorlake). So it seems like not much and that assert() can be left as is.

**_What do you think: should I left that assert() as is, or remove it, or replace with if (...) { log_err(...) } anyway?_**

Few lines above there is another assert: assert(src_info); Which is also always true unless someone introduce a bug. But for this assert() I agree it is better to remove it as in the worst case src_info == NULL we will get a proper fatal handling when trying to access src_info and no magic issues.

Reply:
Don't use asserts alone as an error, because the compiler is allowed to remove them. If it's important that the firmware errors and recovers (which it probably is) then proper error handling should be added in addition to the assert. If the assert can be triggered by invalid inputs (rather than _just_ programming errors) it's invalid as an assert.

Reply:
Still, what should be done if the assert is gone? Do something else?